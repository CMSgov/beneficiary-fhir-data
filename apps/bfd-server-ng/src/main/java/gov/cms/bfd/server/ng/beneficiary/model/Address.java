package gov.cms.bfd.server.ng.beneficiary.model;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import java.util.Optional;

@Embeddable
public class Address {
  @Column(name = "geo_usps_state_cd")
  private String stateCode;

  @Column(name = "geo_zip5_cd")
  private String zipCode;

  @Column(name = "geo_zip_plc_name")
  private String city;

  @Column(name = "bene_line_1_adr")
  private Optional<String> addressLine1;

  @Column(name = "bene_line_2_adr")
  private Optional<String> addressLine2;

  @Column(name = "bene_line_3_adr")
  private Optional<String> addressLine3;

  @Column(name = "bene_line_4_adr")
  private Optional<String> addressLine4;

  @Column(name = "bene_line_5_adr")
  private Optional<String> addressLine5;

  @Column(name = "bene_line_6_adr")
  private Optional<String> addressLine6;

  org.hl7.fhir.r4.model.Address toFhir() {
    var address =
        new org.hl7.fhir.r4.model.Address()
            .setState(stateCode)
            .setPostalCode(zipCode)
            .setCity(city);
    addressLine1.ifPresent(address::addLine);
    addressLine2.ifPresent(address::addLine);
    addressLine3.ifPresent(address::addLine);
    addressLine4.ifPresent(address::addLine);
    addressLine5.ifPresent(address::addLine);
    addressLine6.ifPresent(address::addLine);

    return address;
  }
}
