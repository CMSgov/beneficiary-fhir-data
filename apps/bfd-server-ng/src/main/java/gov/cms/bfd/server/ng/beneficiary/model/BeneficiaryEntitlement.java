package gov.cms.bfd.server.ng.beneficiary.model;

import jakarta.persistence.*;
import java.time.LocalDate;
import java.time.ZonedDateTime;
import lombok.Getter;

/** Entity representing BeneficiaryEntitlement. */
@Entity
@Getter
@Table(name = "beneficiary_entitlement", schema = "idr")
public class BeneficiaryEntitlement {

  @Id
  @Column(name = "bene_sk", nullable = false)
  private Long beneSk;

  @Column(name = "bene_rng_bgn_dt", nullable = false)
  private LocalDate benefitRangeBeginDate;

  @Column(name = "bene_rng_end_dt", nullable = false)
  private LocalDate benefitRangeEndDate;

  @Column(name = "bene_mdcr_entlmt_type_cd", length = 1)
  private String medicareEntitlementTypeCode;

  @Column(name = "bene_mdcr_enrlmt_rsn_cd", length = 1)
  private String medicareEnrollmentReasonCode;

  @Column(name = "bene_mdcr_entlmt_stus_cd", length = 1)
  private String medicareEntitlementStatusCode;

  @Column(name = "idr_trans_efctv_ts", nullable = false)
  private ZonedDateTime idrTransEffectiveTimestamp;

  @Column(name = "idr_trans_obslt_ts", nullable = false)
  private ZonedDateTime idrTransObsoleteTimestamp;

  @Column(name = "idr_updt_ts", nullable = false)
  private ZonedDateTime idrUpdateTimestamp;

  @Column(name = "bfd_created_ts", nullable = false)
  private ZonedDateTime bfdCreatedTimestamp;

  @Column(name = "bfd_updated_ts", nullable = false)
  private ZonedDateTime bfdUpdatedTimestamp;
}
