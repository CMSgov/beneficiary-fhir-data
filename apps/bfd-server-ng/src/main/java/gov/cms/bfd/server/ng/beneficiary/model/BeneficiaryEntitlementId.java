package gov.cms.bfd.server.ng.beneficiary.model;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import java.io.Serializable;
import java.time.LocalDate;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;

/** Represents the composite primary key for the {@link BeneficiaryEntitlement} entity. */
@EqualsAndHashCode
@NoArgsConstructor
@Getter
@AllArgsConstructor
@Embeddable
public class BeneficiaryEntitlementId implements Serializable {

  @Column(name = "bene_sk")
  private Long beneSk;

  @Column(name = "bene_rng_bgn_dt")
  private LocalDate benefitRangeBeginDate;

  @Column(name = "bene_rng_end_dt")
  private LocalDate benefitRangeEndDate;

  @Column(name = "bene_mdcr_entlmt_type_cd")
  private String medicareEntitlementTypeCode;
}
