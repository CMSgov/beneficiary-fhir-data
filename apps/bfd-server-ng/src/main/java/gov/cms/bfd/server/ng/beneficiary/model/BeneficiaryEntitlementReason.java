package gov.cms.bfd.server.ng.beneficiary.model;

import jakarta.persistence.*;
import java.time.LocalDate;
import java.time.ZonedDateTime;
import lombok.Getter;

/** Entity representing BeneficiaryEntitlementReason. */
@Entity
@Getter
@Table(name = "beneficiary_entitlement_reason", schema = "idr")
public class BeneficiaryEntitlementReason {

  @Id
  @Column(name = "bene_sk", nullable = false)
  private Long beneSk;

  @Column(name = "bene_rng_bgn_dt", nullable = false)
  private LocalDate benefitRangeBeginDate;

  @Column(name = "bene_rng_end_dt", nullable = false)
  private LocalDate benefitRangeEndDate;

  @Column(name = "idr_trans_efctv_ts", nullable = false)
  private ZonedDateTime idrTransEffectiveTimestamp;

  @Column(name = "bene_mdcr_entlmt_rsn_cd", length = 1)
  private String medicareEntitlementReasonCode;

  @Column(name = "idr_trans_obslt_ts", nullable = false)
  private ZonedDateTime idrTransObsoleteTimestamp;

  @Column(name = "idr_updt_ts")
  private ZonedDateTime idrUpdateTimestamp;

  @Column(name = "bfd_created_ts", nullable = false)
  private ZonedDateTime bfdCreatedTimestamp;

  @Column(name = "bfd_updated_ts", nullable = false)
  private ZonedDateTime bfdUpdatedTimestamp;
}
