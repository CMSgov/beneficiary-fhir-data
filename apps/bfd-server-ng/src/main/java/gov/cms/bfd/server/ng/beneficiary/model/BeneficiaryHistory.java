package gov.cms.bfd.server.ng.beneficiary.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.util.Optional;

@Entity
@Table(name = "beneficiary_history", schema = "idr")
public class BeneficiaryHistory {
  @Id
  @Column(name = "bene_sk")
  private long beneSk;

  @Column(name = "bene_xref_efctv_sk_computed")
  private long xrefSk;

  @Column(name = "bene_mbi_id")
  private Optional<String> mbi;
}
