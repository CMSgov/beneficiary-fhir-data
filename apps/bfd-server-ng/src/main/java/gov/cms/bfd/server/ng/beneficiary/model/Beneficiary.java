package gov.cms.bfd.server.ng.beneficiary.model;

import gov.cms.bfd.server.ng.DateUtil;
import jakarta.persistence.Column;
import jakarta.persistence.Embedded;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import org.hl7.fhir.r4.model.Patient;

/** Beneficiary table representation. */
@Entity
@Table(name = "beneficiary", schema = "idr")
public class Beneficiary {
  @Id
  @Column(name = "bene_sk", nullable = false)
  private long beneSk;

  @Column(name = "bene_xref_efctv_sk_computed", nullable = false)
  private long xrefSk;

  @Column(name = "bene_mbi_id", nullable = false)
  private String mbi;

  @Column(name = "bene_brth_dt", nullable = false)
  private LocalDate birthDate;

  @Column(name = "bene_race_cd", nullable = false)
  private RaceCode raceCode;

  @Column(name = "bene_sex_cd")
  private Optional<SexCode> sexCode;

  @Column(name = "cntct_lang_cd", nullable = false)
  private LanguageCode languageCode;

  @Embedded private Name beneficiaryName;
  @Embedded private Address address;
  @Embedded private Meta meta;
  @Embedded private DeathDate deathDate;

  /**
   * Transforms the beneficiary record to its FHIR representation.
   *
   * @return patient record
   */
  public Patient toFhir() {
    var patient = new Patient();
    patient.setId(String.valueOf(beneSk));

    patient.setName(List.of(beneficiaryName.toFhir()));
    patient.setBirthDate(DateUtil.toDate(birthDate));
    address.toFhir().ifPresent(a -> patient.setAddress(List.of(a)));
    sexCode.ifPresent(
        s -> {
          patient.setGender(s.toFhirAdministrativeGender());
          patient.addExtension(s.toFhirSexExtension());
        });

    patient.setCommunication(List.of(languageCode.toFhir()));
    deathDate.toFhir().ifPresent(patient::setDeceased);
    patient.addExtension(raceCode.toFhir());
    patient.setMeta(meta.toFhir());

    return patient;
  }
}
