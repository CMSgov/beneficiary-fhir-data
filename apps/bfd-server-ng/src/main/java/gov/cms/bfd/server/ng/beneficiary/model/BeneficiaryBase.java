package gov.cms.bfd.server.ng.beneficiary.model;

import gov.cms.bfd.server.ng.util.DateUtil;
import gov.cms.bfd.server.ng.util.SystemUrls;
import jakarta.persistence.Column;
import jakarta.persistence.Embedded;
import jakarta.persistence.Id;
import jakarta.persistence.MappedSuperclass;
import jakarta.persistence.Transient;
import java.time.LocalDate;
import java.time.ZonedDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import lombok.Getter;
import lombok.NonNull;
import org.hl7.fhir.r4.model.Patient;

/**
 * Base class for beneficiary entities. This is used to prevent excessive fetching for patient
 * information.
 */
@Getter
@MappedSuperclass
public abstract class BeneficiaryBase {
  @Id
  @Column(name = "bene_sk")
  protected long beneSk;

  @Column(name = "bene_xref_efctv_sk_computed")
  protected long xrefSk;

  @Column(name = "bene_brth_dt")
  protected LocalDate birthDate;

  @Column(name = "bene_race_cd")
  protected RaceCode raceCode;

  @Column(name = "bene_sex_cd")
  protected Optional<SexCode> sexCode;

  @Column(name = "cntct_lang_cd")
  protected LanguageCode languageCode;

  @Column(name = "idr_trans_obslt_ts")
  protected ZonedDateTime obsoleteTimestamp;

  @Embedded protected Name beneficiaryName;
  @Embedded protected Address address;
  @Embedded protected Meta meta;
  @Embedded protected DeathDate deathDate;
  @Embedded protected CurrentIdentifier identifier;
  @Transient protected String id = UUID.randomUUID().toString();

  /**
   * Determines if this beneficiary has been merged into another.
   *
   * @return whether the beneficiary is merged
   */
  public boolean isMergedBeneficiary() {
    return beneSk != xrefSk;
  }

  /**
   * Convenience method to convert to FHIR Patient with a specific profile.
   *
   * @param profile the FHIR profile to apply
   * @return patient record
   */
  public Patient toFhir(String profile) {
    var patient = new Patient();

    if (profile.equals(SystemUrls.PROFILE_C4DIC_PATIENT)) {
      patient.setId(id);
    } else {
      patient.setId(String.valueOf(beneSk));
    }

    // Only return a skeleton resource or merged beneficiaries
    if (isMergedBeneficiary()) {
      return patient;
    }

    patient.setCommunication(List.of(languageCode.toFhir()));
    deathDate.toFhir().ifPresent(patient::setDeceased);
    patient.setName(List.of(beneficiaryName.toFhir()));
    patient.setBirthDate(DateUtil.toDate(birthDate));
    address.toFhir().ifPresent(a -> patient.setAddress(List.of(a)));
    sexCode.ifPresent(
            s -> {
              patient.setGender(s.toFhirAdministrativeGender());
              patient.addExtension(s.toFhirSexExtension());
            });
    patient.addExtension(raceCode.toFhir());
    patient.setMeta(meta.toFhirPatient(profile));

    return patient;
  }
}
