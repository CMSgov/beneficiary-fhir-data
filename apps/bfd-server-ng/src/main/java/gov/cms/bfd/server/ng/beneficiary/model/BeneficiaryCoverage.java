package gov.cms.bfd.server.ng.beneficiary.model;

import gov.cms.bfd.server.ng.input.CoverageCompositeId;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import java.util.Collection;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Stream;
import lombok.Getter;
import org.hl7.fhir.r4.model.Coverage;
import org.hl7.fhir.r4.model.Organization;
import org.hl7.fhir.r4.model.Reference;

/** Entity representing the beneficiary table with coverage info. */
@Entity
@Getter
@Table(name = "beneficiary", schema = "idr")
public class BeneficiaryCoverage extends BeneficiaryBase {
  @OneToMany(fetch = FetchType.EAGER)
  @JoinColumn(name = "bene_sk")
  private Set<BeneficiaryEntitlement> beneficiaryEntitlements;

  @OneToMany(fetch = FetchType.EAGER)
  @JoinColumn(name = "bene_sk")
  private Set<BeneficiaryThirdParty> beneficiaryThirdParties;

  @OneToOne(fetch = FetchType.EAGER)
  @JoinColumn(name = "bene_sk")
  private BeneficiaryStatus beneficiaryStatus;

  @OneToOne(fetch = FetchType.EAGER)
  @JoinColumn(name = "bene_sk")
  private BeneficiaryEntitlementReason beneficiaryEntitlementReason;

  private Optional<BeneficiaryEntitlementReason> getEntitlementReason() {
    return Optional.ofNullable(beneficiaryEntitlementReason);
  }

  private Optional<BeneficiaryStatus> getStatus() {
    return Optional.ofNullable(beneficiaryStatus);
  }

  /**
   * Creates a FHIR Coverage resource.
   *
   * @param coverageCompositeId The full ID for the Coverage resource.
   * @return A FHIR Coverage object.
   */
  public Coverage toFhir(CoverageCompositeId coverageCompositeId) {
    var coverage = new Coverage();
    coverage.setId(coverageCompositeId.fullId());

    var coveragePart = coverageCompositeId.coveragePart();
    var coverageType = coveragePart.getStandardCode();
    var entitlementOpt =
        beneficiaryEntitlements.stream()
            .filter(
                e -> e.getId().getMedicareEntitlementTypeCode().toUpperCase().equals(coverageType))
            .findFirst();
    // If no entitlement record, bene doesn't have this coverage type
    if (entitlementOpt.isEmpty()) {
      return coverage;
    }
    var entitlement = entitlementOpt.get();

    coverage.setMeta(meta.toFhirCoverage());

    coverage.setBeneficiary(new Reference("Patient/" + beneSk));

    coverage.setRelationship(RelationshipFactory.createSelfSubscriberRelationship());

    Organization cmsOrg = OrganizationFactory.createCmsOrganization();
    coverage.addContained(cmsOrg);

    coverage.addPayor(new Reference().setReference("#" + cmsOrg.getIdElement().getIdPart()));

    identity.toFhirMbiIdentifier().ifPresent(coverage::addIdentifier);
    coverage.setSubscriberId(identity.getMbiValue());

    coverage.setType(coveragePart.toFhirTypeCode());
    coverage.addClass_(coveragePart.toFhirClassComponent());
    var fhirPeriodOpt = entitlement.toFhirPeriod();
    fhirPeriodOpt.ifPresent(coverage::setPeriod);
    coverage.setStatus(Coverage.CoverageStatus.ACTIVE);

    var beneficiaryThirdParty =
        beneficiaryThirdParties.stream()
            .filter(tp -> tp.getId().getThirdPartyTypeCode().toUpperCase().equals(coverageType))
            .findFirst();

    Stream.of(
            beneficiaryThirdParty.map(BeneficiaryThirdParty::toFhirExtensions),
            Optional.of(entitlement.toFhirExtensions()),
            getStatus().map(BeneficiaryStatus::toFhirExtensions),
            getEntitlementReason().map(BeneficiaryEntitlementReason::toFhirExtensions))
        .flatMap(Optional::stream)
        .flatMap(Collection::stream)
        .forEach(coverage::addExtension);

    return coverage;
  }

  /**
   * Creates a FHIR Coverage resource or None if the beneficiary does not have the matching coverage
   * type.
   *
   * @param coverageCompositeId The full ID for the Coverage resource.
   * @return A FHIR Coverage object.
   */
  public Optional<Coverage> toFhirCoverageIfPresent(CoverageCompositeId coverageCompositeId) {
    var coverage = toFhir(coverageCompositeId);
    if (coverage.getIdentifier().isEmpty()) {
      return Optional.empty();
    }
    return Optional.of(coverage);
  }
}
