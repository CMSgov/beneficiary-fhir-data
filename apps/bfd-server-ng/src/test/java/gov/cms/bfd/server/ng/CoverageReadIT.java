package gov.cms.bfd.server.ng;

import static org.hamcrest.Matchers.containsString;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import ca.uhn.fhir.rest.gclient.IReadTyped;
import ca.uhn.fhir.rest.server.exceptions.InvalidRequestException;
import ca.uhn.fhir.rest.server.exceptions.ResourceNotFoundException;
import io.restassured.RestAssured;
import org.hl7.fhir.r4.model.Coverage;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EmptySource;
import org.junit.jupiter.params.provider.ValueSource;

public class CoverageReadIT extends IntegrationTestBase {

  private static final String BENE_ID_PART_A_ONLY = "178083966";
  private static final String BENE_ID_PART_B_ONLY = "365359727";
  private static final String BENE_ID_HAS_A_AND_B = "405764107";
  private static final String BENE_ID_NO_TP = "451482106";
  private static final String BENE_ID_EXPIRED_COVERAGE = "421056595";
  private static final String BENE_ID_FUTURE_COVERAGE = "971050241";

  private IReadTyped<Coverage> coverageRead() {
    return getFhirClient().read().resource(Coverage.class);
  }

  private String createCoverageId(String part, String beneId) {
    return String.format("part-%s-%s", part, beneId);
  }

  @Test
  void coverageReadValidPartACompositeId() {
    var validCoverageId = createCoverageId("a", BENE_ID_PART_A_ONLY);

    var coverage = coverageRead().withId(validCoverageId).execute();
    assertNotNull(coverage, "Coverage resource should not be null for a valid ID");
    assertFalse(coverage.isEmpty(), "Coverage resource should not be empty for a valid ID");
    expect.scenario("validPartA").serializer("fhir+json").toMatchSnapshot(coverage);

    var missingCoverageId = createCoverageId("b", BENE_ID_PART_A_ONLY);
    var missingCoverage = coverageRead().withId(missingCoverageId).execute();
    // Response for a missing coverage part should only contain the ID
    assertTrue(missingCoverage.getIdentifier().isEmpty());
    assertNotNull(missingCoverage, "Coverage resource should not be null for a valid ID");
    assertFalse(missingCoverage.isEmpty(), "Coverage resource should not be empty for a valid ID");
    expect.scenario("missingPartB").serializer("fhir+json").toMatchSnapshot(missingCoverage);
  }

  @Test
  void coverageReadValidPartBCompositeId() {
    var validCoverageId = createCoverageId("b", BENE_ID_PART_B_ONLY);

    var coverage = coverageRead().withId(validCoverageId).execute();
    assertNotNull(coverage, "Coverage resource should not be null for a valid ID");
    assertFalse(coverage.isEmpty(), "Coverage resource should not be empty for a valid ID");
    expect.scenario("validPartB").serializer("fhir+json").toMatchSnapshot(coverage);

    var missingCoverageId = createCoverageId("a", BENE_ID_PART_B_ONLY);
    var missingCoverage = coverageRead().withId(missingCoverageId).execute();
    // Response for a missing coverage part should only contain the ID
    assertTrue(missingCoverage.getIdentifier().isEmpty());
    assertNotNull(missingCoverage, "Coverage resource should not be null for a valid ID");
    assertFalse(missingCoverage.isEmpty(), "Coverage resource should not be empty for a valid ID");
    expect.scenario("missingPartA").serializer("fhir+json").toMatchSnapshot(missingCoverage);
  }

  @Test
  void coverageReadForNonCurrentEffectiveBeneficiaryIdShouldBeNotFound() {

    var nonCurrentEffectiveBeneId = "part-a-181968400";
    assertThrows(
        ResourceNotFoundException.class,
        () -> coverageRead().withId(nonCurrentEffectiveBeneId).execute(),
        "Should throw ResourceNotFoundException for Coverage ID 'part-a-181968400' "
            + "because the beneficiary record is not the current effective version.");
  }

  @Test
  void coverageReadValidCompositeId() {
    var validCoverageId = createCoverageId("b", BENE_ID_HAS_A_AND_B);

    var coverage = coverageRead().withId(validCoverageId).execute();
    assertNotNull(coverage, "Coverage resource should not be null for a valid ID");
    assertFalse(coverage.isEmpty(), "Coverage resource should not be empty for a valid ID");
  }

  @Test
  void coverageReadBeneExistsButPartOrVersionNotFound() {
    var idForNonExistentCoverage = "part-c-1";

    assertThrows(
        InvalidRequestException.class,
        () -> coverageRead().withId(idForNonExistentCoverage).execute(),
        "Should throw ResourceNotFoundException for a validly formatted ID that doesn't map to a resource.");
  }

  @Test
  void coverageReadBeneSkNotFound() {
    var nonExistentBeneSkId = "part-a-9999999";

    assertThrows(
        ResourceNotFoundException.class,
        () -> coverageRead().withId(nonExistentBeneSkId).execute(),
        "Should throw ResourceNotFoundException if the beneficiary part of the ID does not exist.");
  }

  @ParameterizedTest
  @ValueSource(strings = {"part-a", "-12345", "part-a-abc", "foo-12345", "part-e-12345"})
  void coverageReadInvalidIdFormatBadRequest(String invalidId) {
    assertThrows(
        InvalidRequestException.class,
        () -> coverageRead().withId(invalidId).execute(),
        "Should throw InvalidRequestException from server for ID: " + invalidId);
  }

  @Test
  void coverageReadUnsupportedValidPartBadRequest() {
    var unsupportedPartId = "part-c-1";
    assertThrows(
        InvalidRequestException.class,
        () -> coverageRead().withId(unsupportedPartId).execute(),
        "Should throw InvalidRequestException for an unsupported part like Part C/D.");
  }

  @Test
  void coverageReadPartACoverageNotFound() {
    var expiredCoverageId = createCoverageId("a", "848484848");

    assertThrows(
        ResourceNotFoundException.class,
        () -> coverageRead().withId(expiredCoverageId).execute(),
        "Should throw ResourceNotFoundException.");
  }

  @ParameterizedTest
  @ValueSource(strings = {"A", "B"})
  void coverageReadBeneWithNoCoverageReturnsEmpty(String part) {
    final String partId = createCoverageId(part, "289169129");
    var coverage = coverageRead().withId(partId).execute();
    assertEquals(partId.toLowerCase(), coverage.getIdPart());
    assertTrue(coverage.getIdentifier().isEmpty());
    expect.scenario("missingBothPart" + part).serializer("fhir+json").toMatchSnapshot(coverage);
  }

  @Test
  void coverageReadForBeneWithOnlyPastEntitlementPeriodsShouldBeEmpty() {
    final var partBId = createCoverageId("b", BENE_ID_EXPIRED_COVERAGE);

    var coverage = coverageRead().withId(partBId).execute();
    assertEquals(partBId, coverage.getIdPart());
    assertTrue(coverage.getIdentifier().isEmpty());
    expect.serializer("fhir+json").toMatchSnapshot(coverage);
  }

  @Test
  void coverageReadForBeneWithOnlyFutureEntitlementPeriodsShouldBeEmpty() {
    final String partBId = createCoverageId("b", BENE_ID_FUTURE_COVERAGE);

    var coverage = coverageRead().withId(partBId).execute();
    assertEquals(partBId, coverage.getIdPart());
    assertTrue(coverage.getIdentifier().isEmpty());
    expect.serializer("fhir+json").toMatchSnapshot(coverage);
  }

  @Test
  void coverageRead_forBeneWithMissingTpDataShouldStillReturnResource() {

    var partAId = createCoverageId("a", BENE_ID_NO_TP);
    var coverage = coverageRead().withId(partAId).execute();

    assertNotNull(coverage, "Coverage resource should not be null even without TPL data.");
    assertFalse(coverage.isEmpty(), "Coverage resource should not be empty.");

    expect.serializer("fhir+json").toMatchSnapshot(coverage);
  }

  @ParameterizedTest
  @EmptySource
  void coverageReadEmptyIdSegmentResultsInServerError(String id) {
    RestAssured.given()
        .when()
        .get(getServerUrl() + "/Coverage/" + id)
        .then()
        .statusCode(400)
        .body(
            "issue[0].diagnostics",
            containsString(
                "The FHIR endpoint on this server does not know how to handle GET operation[Coverage/] with parameters [[]]"));
  }
}
