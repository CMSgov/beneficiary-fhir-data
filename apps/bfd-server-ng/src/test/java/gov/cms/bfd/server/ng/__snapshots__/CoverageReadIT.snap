gov.cms.bfd.server.ng.CoverageReadIT.coverageReadValidPartACompositeId=[
{
  "beneficiary" : {
    "reference" : "Patient/1"
  },
  "class" : [ {
    "type" : {
      "coding" : [ {
        "code" : "plan",
        "system" : "http://terminology.hl7.org/CodeSystem/coverage-class"
      } ]
    },
    "value" : "Part A"
  } ],
  "contained" : [ {
    "active" : true,
    "contact" : [ {
      "purpose" : {
        "coding" : [ {
          "code" : "PATINF",
          "display" : "Patient",
          "system" : "http://terminology.hl7.org/CodeSystem/contactentity-type"
        } ]
      },
      "telecom" : [ {
        "system" : "phone",
        "value" : "1-800-633-4227"
      }, {
        "system" : "phone",
        "value" : "TTY: 1-877-486-2048"
      }, {
        "system" : "url",
        "value" : "www.medicare.gov"
      } ]
    } ],
    "id" : "cms-org",
    "meta" : {
      "profile" : [ "http://hl7.org/fhir/us/carin-bb/StructureDefinition/C4BB-Organization|2.1.0" ]
    },
    "name" : "Centers for Medicare and Medicaid Services",
    "resourceType" : "Organization"
  } ],
  "id" : "part-a-1",
  "identifier" : [ {
    "system" : "http://hl7.org/fhir/sid/us-mbi",
    "type" : {
      "coding" : [ {
        "code" : "MB",
        "system" : "http://terminology.hl7.org/CodeSystem/v2-0203"
      } ]
    },
    "value" : "1S000000000"
  } ],
  "meta" : {
    "lastUpdated" : "2024-01-01T00:00:00.000+00:00",
    "profile" : [ "http://hl7.org/fhir/us/carin-bb/StructureDefinition/C4BB-Coverage|2.1.0", "http://hl7.org/fhir/us/core/StructureDefinition/us-core-coverage|6.1.0" ]
  },
  "payor" : [ {
    "reference" : "#cms-org"
  } ],
  "relationship" : {
    "coding" : [ {
      "code" : "self",
      "display" : "Self",
      "system" : "http://terminology.hl7.org/CodeSystem/subscriber-relationship"
    } ]
  },
  "resourceType" : "Coverage",
  "subscriberId" : "1S000000000",
  "type" : {
    "coding" : [ {
      "code" : "1",
      "display" : "MEDICARE",
      "system" : "https://nahdo.org/sopt"
    } ]
  }
}
]


gov.cms.bfd.server.ng.CoverageReadIT.coverageReadValidPartBCompositeId=[
{
  "beneficiary" : {
    "reference" : "Patient/1"
  },
  "class" : [ {
    "type" : {
      "coding" : [ {
        "code" : "plan",
        "system" : "http://terminology.hl7.org/CodeSystem/coverage-class"
      } ]
    },
    "value" : "Part B"
  } ],
  "contained" : [ {
    "active" : true,
    "contact" : [ {
      "purpose" : {
        "coding" : [ {
          "code" : "PATINF",
          "display" : "Patient",
          "system" : "http://terminology.hl7.org/CodeSystem/contactentity-type"
        } ]
      },
      "telecom" : [ {
        "system" : "phone",
        "value" : "1-800-633-4227"
      }, {
        "system" : "phone",
        "value" : "TTY: 1-877-486-2048"
      }, {
        "system" : "url",
        "value" : "www.medicare.gov"
      } ]
    } ],
    "id" : "cms-org",
    "meta" : {
      "profile" : [ "http://hl7.org/fhir/us/carin-bb/StructureDefinition/C4BB-Organization|2.1.0" ]
    },
    "name" : "Centers for Medicare and Medicaid Services",
    "resourceType" : "Organization"
  } ],
  "id" : "part-b-1",
  "identifier" : [ {
    "system" : "http://hl7.org/fhir/sid/us-mbi",
    "type" : {
      "coding" : [ {
        "code" : "MB",
        "system" : "http://terminology.hl7.org/CodeSystem/v2-0203"
      } ]
    },
    "value" : "1S000000000"
  } ],
  "meta" : {
    "lastUpdated" : "2024-01-01T00:00:00.000+00:00",
    "profile" : [ "http://hl7.org/fhir/us/carin-bb/StructureDefinition/C4BB-Coverage|2.1.0", "http://hl7.org/fhir/us/core/StructureDefinition/us-core-coverage|6.1.0" ]
  },
  "payor" : [ {
    "reference" : "#cms-org"
  } ],
  "relationship" : {
    "coding" : [ {
      "code" : "self",
      "display" : "Self",
      "system" : "http://terminology.hl7.org/CodeSystem/subscriber-relationship"
    } ]
  },
  "resourceType" : "Coverage",
  "subscriberId" : "1S000000000",
  "type" : {
    "coding" : [ {
      "code" : "121",
      "display" : "MEDICARE FFS",
      "system" : "https://nahdo.org/sopt"
    } ]
  }
}
]